File: services/go-clean/internal/repo/health_probes.go
package repo

import (
	"context"
	"database/sql"

	"github.com/aws/aws-sdk-go-v2/service/cognitoidentityprovider"
	"github.com/aws/aws-sdk-go-v2/service/s3"
)

// DBProbe は MySQL の疎通確認。
type DBProbe struct{ db *sql.DB }

func NewDBProbe(db *sql.DB) *DBProbe { return &DBProbe{db: db} }

func (p *DBProbe) Name() string { return "database" }

func (p *DBProbe) Check(ctx context.Context) error {
	return p.db.PingContext(ctx)
}

// StorageProbe は S3（本番）/ SeaweedFS（ローカル）の疎通確認。
//
// endpoint を差し替えるだけで両方に対応する。
// S3 互換 API を使う限り、コードは共通で済む。
type StorageProbe struct {
	client *s3.Client
	bucket string
}

func NewStorageProbe(client *s3.Client, bucket string) *StorageProbe {
	return &StorageProbe{client: client, bucket: bucket}
}

func (p *StorageProbe) Name() string { return "object_storage" }

func (p *StorageProbe) Check(ctx context.Context) error {
	// オブジェクト一覧ではなく HeadBucket を使う。
	// 必要な権限が最小で済み、バケットの中身の量に影響されない。
	_, err := p.client.HeadBucket(ctx, &s3.HeadBucketInput{Bucket: &p.bucket})
	return err
}

// CognitoProbe は Cognito の疎通確認。
type CognitoProbe struct {
	client     *cognitoidentityprovider.Client
	userPoolID string
}

func NewCognitoProbe(client *cognitoidentityprovider.Client, userPoolID string) *CognitoProbe {
	return &CognitoProbe{client: client, userPoolID: userPoolID}
}

func (p *CognitoProbe) Name() string { return "cognito" }

func (p *CognitoProbe) Check(ctx context.Context) error {
	_, err := p.client.DescribeUserPool(ctx, &cognitoidentityprovider.DescribeUserPoolInput{
		UserPoolId: &p.userPoolID,
	})
	return err
}

File: services/dotnet-clean/src/Infrastructure/Health/HealthProbes.cs
using Amazon.CognitoIdentityProvider;
using Amazon.CognitoIdentityProvider.Model;
using Amazon.S3;
using Amazon.S3.Model;
using Application.Abstractions;
using Infrastructure.Cognito;
using Infrastructure.Persistence;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Options;

namespace Infrastructure.Health;

/// <summary>MySQL の疎通確認。</summary>
internal sealed class DatabaseProbe(AppDbContext db) : IHealthProbe
{
    public string Name => "database";

    public async Task CheckAsync(CancellationToken cancellationToken)
    {
        // CanConnectAsync は bool を返し、失敗理由を捨ててしまう。
        // 実際にクエリを投げて、例外をそのまま上へ渡す。
        await db.Database.ExecuteSqlRawAsync("SELECT 1", cancellationToken);
    }
}

/// <summary>
/// S3（本番）/ SeaweedFS（ローカル）の疎通確認。
///
/// endpoint を差し替えるだけで両方に対応する。
/// S3 互換 API を使う限り、コードは共通で済む。
/// </summary>
internal sealed class ObjectStorageProbe(IAmazonS3 s3, IOptions<StorageOptions> options) : IHealthProbe
{
    public string Name => "object_storage";

    public async Task CheckAsync(CancellationToken cancellationToken)
    {
        // オブジェクト一覧ではなく HeadBucket 相当を使う。
        // 必要な権限が最小で済み、バケットの中身の量に影響されない。
        await s3.GetBucketLocationAsync(
            new GetBucketLocationRequest { BucketName = options.Value.Bucket },
            cancellationToken);
    }
}

/// <summary>Cognito の疎通確認。</summary>
internal sealed class CognitoProbe(
    IAmazonCognitoIdentityProvider client,
    IOptions<CognitoOptions> options) : IHealthProbe
{
    public string Name => "cognito";

    public async Task CheckAsync(CancellationToken cancellationToken)
    {
        await client.DescribeUserPoolAsync(
            new DescribeUserPoolRequest { UserPoolId = options.Value.UserPoolId },
            cancellationToken);
    }
}

/// <summary>オブジェクトストレージの設定。</summary>
public sealed class StorageOptions
{
    public string Bucket { get; set; } = "app-static";

    /// <summary>ローカルの S3 互換ストレージのエンドポイント。本番では空。</summary>
    public string Endpoint { get; set; } = string.Empty;
}
